### PR TITLE
Pyclasses2

### PR DESCRIPTION
**Pull request by jsalvatier**

### CODE REVIEW

**Comment by jsalvatier** [comment]
Maybe has stuff that needs to be cleaned up, but the tests are finally passing.

File: pymc/model.py [hunk 1]
```diff
@@ -40,28 +42,65 @@ def modelcontext(model):
         return Model.get_context()
     return model
 
+class Factor(object):
```
Comment (fonnesbeck) [inline pymc/model.py:48]: All the classes in model.py should probably have docstrings.

File: pymc/distributions/transforms.py [hunk 2]
```diff
@@ -1,42 +1,49 @@
...
-def transform(name, forward, backward, jacobian_det, getshape=None):
-    def apply(dist):
 
-        @quickclass(TensorDist)
-        def TransformedDistribtuion():
-            try:
-                testval = forward(dist.testval)
-            except TypeError:
-                testval = dist.testval
+class Transform(object):
```
Comment (fonnesbeck) [inline pymc/distributions/transforms.py:6]: The classes and functions in transforms.py should all have docstrings.

File: pymc/model.py (hunk 1 above)
Comment (jsalvatier) [inline pymc/model.py:48]: Done.

File: pymc/data.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import pkgutil
+import io
+
+__all__ = ['get_data_file']
+
+def get_data_file(pkg, path):
```
Comment (jsalvatier) [inline pymc/data.py:6]: Needs docstring.

File: pymc/distributions/continuous.py [hunk 4]
```diff
@@ -406,36 +363,38 @@ def logp(value):
             alpha > 0,
             beta > 0)
 
-    logp.__doc__ = """
-        Gamma log-likelihood with paramters alpha={0} and beta={1}.
+class Bounded(Continuous):
```
Comment (jsalvatier) [inline pymc/distributions/continuous.py:366]: Docstring.

File: pymc/distributions/continuous.py [hunk 5]
```diff
@@ -406,36 +363,38 @@ def logp(value):
...
-            return bound(
-                dist.logp(value),
 
-                lower <= value, value <= upper)
 
-        if hasattr(dist, 'mode'):
-            mode = dist.mode
+class Bound(object):
+    def __init__(self, distribution, lower=-inf, upper=inf):
+        """A bounded distribution."""
```
Comment (jsalvatier) [inline pymc/distributions/continuous.py]: Docstring on the class.

File: pymc/distributions/distribution.py [hunk 6]
```diff
@@ -1,11 +1,8 @@
...
-from ..model import *
-import warnings
+from ..model import Model
 
-__all__ = ['DensityDist', 'TensorDist', 'tensordist', 'continuous',
-           'discrete', 'arbitrary', 'evaluate']
+__all__ = ['DensityDist', 'Distribution', 'Arbitrary', 'Continuous', 'Discrete']
 
 
 class Distribution(object):
```
Comment (jsalvatier) [inline pymc/distributions/distribution.py:8]: Docstring.

File: pymc/distributions/distribution.py [hunk 7]
```diff
@@ -60,48 +40,45 @@ def default(self, testval):
...
-
+def TensorType(dtype, shape):
+    return t.TensorType(str(dtype), np.atleast_1d(shape) == 1)
 
-def tensordist(defaults):
-    def decorator(fn):
-        fn = withdefaults(defaults)(fn)
-        return quickclass(TensorDist)(fn)
-    return decorator
+class Arbitrary(Distribution):
```
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: Docstring.

File: pymc/distributions/distribution.py [hunk 8]
```diff
@@ -60,48 +40,45 @@ def default(self, testval):
...
+    def __init__(self, shape=(), dtype='float64', *args, **kwargs):
+        Distribution.__init__(self, *args, **kwargs)
 
+        self.shape = np.atleast_1d(shape)
+        self.type = TensorType(dtype, shape)
+        self.__dict__.update(locals())
 
-def TensorType(dtype, shape):
-    return t.TensorType(str(dtype), np.atleast_1d(shape) == 1)
+class Discrete(Distribution):
```
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: Docstring.

File: pymc/distributions/distribution.py [hunk 9]
```diff
@@ -60,48 +40,45 @@ def default(self, testval):
...
-    return t.TensorType(str(dtype), np.atleast_1d(shape) == 1)
+class Discrete(Distribution): 
+    def __init__(self, shape=(), dtype='int64', testval=None, *args, **kwargs):
+        Distribution.__init__(self, *args, **kwargs)
+        self.shape = np.atleast_1d(shape)
+        self.type = TensorType(dtype, shape)
+        self.default_testvals = ['mode']
+        self.__dict__.update(locals())
 
+class Continuous(Distribution):
```
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: Docstring

File: pymc/distributions/distribution.py [hunk 10]
```diff
@@ -60,48 +40,45 @@ def default(self, testval):
...
 
+        self.__dict__.update(locals())
 
-def discrete(shape=(), dtype='int64', testval=None):
-    shape = np.atleast_1d(shape)
-    type = TensorType(dtype, shape)
-    default_testvals = ['mode']
-    return locals()
 
+class DensityDist(Arbitrary):
```
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: Docstring.

File: pymc/step_methods/compound.py [hunk 11]
```diff
@@ -3,15 +3,12 @@
 
 @author: johnsalvatier
 '''
-from ..quickclass import *
 
+class CompoundStep(object):
```
Comment (jsalvatier) [inline pymc/step_methods/compound.py:7]: Docstring

File: pymc/tests/test_examples.py [hunk 12]
```diff
@@ -1,37 +1,28 @@
...
 
+def test_examples1():
+    for t in itertools.islice(get_examples(), 10, 20):
+        yield t
 
-def check_example(p):
-    os.chdir(path.dirname(p))
-    imp.load_source('__main__', path.basename(p))
+def test_examples2():
+    for t in itertools.islice(get_examples(), 20, 30):
```
Comment (jsalvatier) [inline pymc/tests/test_examples.py]: This should include the rest of them so we don't accidentally not run some.

File: pymc/data.py (hunk 3 above)
Comment (jsalvatier) [inline pymc/data.py:6]: done

File: pymc/distributions/distribution.py (hunk 6 above)
Comment (jsalvatier) [inline pymc/distributions/distribution.py:8]: done

File: pymc/distributions/distribution.py (hunk 7 above)
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: removed.

File: pymc/distributions/distribution.py (hunk 9 above)
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: done

File: pymc/distributions/distribution.py (hunk 8 above)
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: done

File: pymc/distributions/distribution.py (hunk 10 above)
Comment (jsalvatier) [inline pymc/distributions/distribution.py]: done

**Comment by fonnesbeck** [comment]
The following error comes up on TravisCI:

```
======================================================================
ERROR: pymc.tests.test_examples.test_examples1('pymc.examples.python_vs_c',)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.3/lib/python3.3/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc/pymc/tests/test_examples.py", line 15, in check_example
    example.run("short")
  File "/home/travis/build/pymc-devs/pymc/pymc/examples/python_vs_c.py", line 51, in run
    mode.print_summary()
  File "/home/travis/virtualenv/python3.3/lib/python3.3/site-packages/theano/compile/profilemode.py", line 256, in print_summary
    **kwargs)
  File "/home/travis/virtualenv/python3.3/lib/python3.3/site-packages/theano/compile/profilemode.py", line 454, in print_summary_
    otimes.sort()
nose.proxy.TypeError: TypeError: unorderable types: DimShuffle() < Alloc()
-------------------- >> begin captured stdout << ---------------------
ProfileMode(linker = py, optimizer = fast_run)
[array(-2779.03706509484), array([ -126.1376121 ,   374.34752551,   581.81564097, -1055.00960427])]
ProfileMode is deprecated! Use the new profiler.
 The Theano flags to enable it ise: profile=True
 The Theano flags for the memory profile to it is: profile_memory=True
```

Running it locally to see if I can replicate.

**Comment by jsalvatier** [comment]
Yeah, I have no idea what that's about. I commented out the call to print_summary for now.

**Comment by jsalvatier** [comment]
I removed it, but looks like even just rerunning it fixed it.

**Comment by fonnesbeck** [comment]
Time to merge this, I think. We can make additional changes after the merge, as necessary.

**Pull request merged**